com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.refactor.cs

```diff
@@ -0,0 +1,262 @@
+                
+                var rootObjects = scene.GetRootGameObjects();
+                var upgradables = GetUpgradables(rootObjects, m_ObjectUpgrader.RootUpgradeComponentTypes, true);
+                UpgradeNonPrefabs(upgradables, upgradedObjects, timelineManager);
+                
+#if CINEMACHINE_TIMELINE
+                var playableDirectors = TimelineManager.GetPlayableDirectors(scene);
+                UpdateAnimationReferences(playableDirectors);
+#endif
+                UpgradeObjectReferences(rootObjects);
+                
+                // Clean up all obsolete components
+                foreach (var go in rootObjects)
+                    m_ObjectUpgrader.DeleteObsoleteComponents(go);
+                
+#if CINEMACHINE_TIMELINE
+                // Restore timeline names
+                foreach (var director in playableDirectors)
+                {
+                    if (timelineRenames.ContainsKey(director.name)) // search based on guid name
+                    {
+                        director.name = timelineRenames[director.name]; // restore director name
+                    }
+                }
+#endif
+                EditorSceneManager.SaveScene(scene);
+            }
+
+            m_CurrentSceneOrPrefab = string.Empty;
+        }
+
+        /// <summary>
+        /// For each prefab asset
+        /// - Delete obsolete components
+        /// - Update manager camera caches
+        /// </summary>
+        void CleanupPrefabAssets()
+        {
+            for (var p = 0; p < m_PrefabManager.PrefabCount; ++p)
+            {
+                m_CurrentSceneOrPrefab = m_PrefabManager.GetPrefabAssetPath(p);
+                using var editingScope = new PrefabUtility.EditPrefabContentsScope(m_CurrentSceneOrPrefab);
+                var prefabContents = editingScope.prefabContentsRoot;
+                var components = new List<Component>();
+                foreach (var type in m_ObjectUpgrader.RootUpgradeComponentTypes)
+                    components.AddRange(prefabContents.GetComponentsInChildren(type, true).ToList());
+                
+                foreach (var c in components)
+                {
+                    if (c == null || c.gameObject == null)
+                        continue; // was a hidden rig
+                    if (c.GetComponentInParent<CinemachineDoNotUpgrade>(true) != null)
+                        continue; // is a backup copy
+
+                    m_ObjectUpgrader.DeleteObsoleteComponents(c.gameObject);
+                }
+                    
+                var managers =
+                    prefabContents.GetComponentsInChildren<CinemachineCameraManagerBase>();
+                foreach (var manager in managers)
+                {
+                    manager.InvalidateCameraCache();
+                    var justToUpdateCache = manager.ChildCameras;
+                }
```

Comment:
Is this needed?  Feels hacky.

Reply:
Not, if we'll add an a code in the inspector to invalidate cache on scene dirty/reload.

Reply:
I'll remove this and add a proper fix in the Editor in a separate PR.